File: VB/TimeEditControl/TimeEditPopupControl/TimeEditPopupControlPainter.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.Skins
Imports System.Drawing
Imports DevExpress.Utils.Drawing

Namespace TimeEditControl

	Public Class TimeEditPopupControlPainter
		Private PopupViewInfo As TimeEditPopupControlViewInfo
		Private drawPen As New Pen(New SolidBrush(Color.Black))

		Public Sub New(ByVal viewInfo As TimeEditPopupControlViewInfo)
			PopupViewInfo = viewInfo
		End Sub

		Protected Overrides Sub Finalize()
			drawPen = Nothing
			PopupViewInfo = Nothing
		End Sub

		Public Sub Draw(ByVal cache As GraphicsCache)
			DrawContent(cache)
		End Sub

		Private Sub DrawContent(ByVal cache As GraphicsCache)
			DrawLabels(cache)
			DrawCells(cache)
		End Sub

		Protected Overridable Sub DrawLabels(ByVal cache As GraphicsCache)
			drawPen.Color =PopupViewInfo.PopupAppearance.BackColor
			Dim HourSize As SizeF = cache.Graphics.MeasureString("Hour", PopupViewInfo.PopupAppearance.Font)
			Dim MinuteSize As SizeF = cache.Graphics.MeasureString("Minute", PopupViewInfo.PopupAppearance.Font)

			Dim HourPoint As PointF = CalcLabelPoint(PopupViewInfo.CellsList(1), HourSize)
			Dim MinutePoint As PointF = CalcLabelPoint(PopupViewInfo.CellsList(13), MinuteSize)

			cache.Graphics.DrawString("Hour", PopupViewInfo.PopupAppearance.Font, drawPen.Brush, HourPoint)
			cache.Graphics.DrawString("Minute", PopupViewInfo.PopupAppearance.Font, drawPen.Brush, MinutePoint)

		End Sub

		Protected Overridable Sub DrawCells(ByVal cache As GraphicsCache)
			For Each cellInfo As CellViewInfo In PopupViewInfo.CellsList
				drawPen.Color = PopupViewInfo.PopupAppearance.BackColor
				If cellInfo.isSelected Then
					ObjectPainter.DrawObject(cache, SkinElementPainter.Default, GetCellElementInfo(cellInfo, cache, True))
					drawPen.Color = PopupViewInfo.PopupAppearance.BackColor2
				End If
				If cellInfo.isHotPointed Then
					ObjectPainter.DrawObject(cache, SkinElementPainter.Default, GetCellElementInfo(cellInfo, cache, False))
					drawPen.Color = PopupViewInfo.PopupAppearance.BackColor2
				End If
				cache.Graphics.DrawString(cellInfo.CellText, cellInfo.cellTextFont,drawPen.Brush, SetTextPoint(cellInfo, cache.Graphics))
			Next cellInfo
		End Sub

		Private Function SetTextPoint(ByVal cellInfo As CellViewInfo, ByVal g As Graphics) As PointF
			Dim textSize As SizeF = g.MeasureString(cellInfo.CellText, PopupViewInfo.PopupAppearance.Font)
			Dim x As Single = cellInfo.cellBounds.X + cellInfo.cellBounds.Width - 3 - textSize.Width
			Dim y As Single = cellInfo.cellBounds.Y + cellInfo.cellBounds.Height - 1 - textSize.Height

			Return New PointF(x, y)
		End Function

		Protected Overridable Function GetCellElementInfo(ByVal cell As CellViewInfo, ByVal cache As GraphicsCache, ByVal selected As Boolean) As SkinElementInfo
			Dim skin As Skin = CommonSkins.GetSkin(PopupViewInfo.ownerControl.OwnerEdit.LookAndFeel.ActiveLookAndFeel)
			Dim skinInfo As SkinElementInfo
			If selected Then
				skinInfo = New SkinElementInfo(skin(CommonSkins.SkinSelection), cell.cellBounds)
			Else
				skinInfo = New SkinElementInfo(skin(CommonSkins.SkinHighlightedItem), cell.cellBounds)
			End If
			skinInfo.Cache = cache
			Return skinInfo
		End Function

		Private Function CalcLabelPoint(ByVal cell As CellViewInfo, ByVal textSize As SizeF) As PointF
			Const indent As Integer = 2
			Dim x As Single = cell.cellBounds.X + cell.cellBounds.Width \ 2 - textSize.Width \ 2
			Dim y As Single = cell.cellBounds.Y - textSize.Height - indent
			Return New PointF(x, y)
		End Function
	End Class
End Namespace

File: CS/TimeEditControl/TimeEditPopupControl/TimeEditPopupControlPainter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DevExpress.Skins;
using System.Drawing;
using DevExpress.Utils.Drawing;

namespace TimeEditControl
{

    public class TimeEditPopupControlPainter
    {
        TimeEditPopupControlViewInfo PopupViewInfo;
        Pen drawPen = new Pen(new SolidBrush(Color.Black));

        public TimeEditPopupControlPainter(TimeEditPopupControlViewInfo viewInfo)
        {
            PopupViewInfo = viewInfo;
        }

        ~TimeEditPopupControlPainter()
        {
            drawPen = null;
            PopupViewInfo = null;
        }

        public void Draw(GraphicsCache cache)
        {
            DrawContent(cache);
        }

        private void DrawContent(GraphicsCache cache)
        {
            DrawLabels(cache);
            DrawCells(cache);
        }

        protected virtual void DrawLabels(GraphicsCache cache)
        {
            drawPen.Color =PopupViewInfo.PopupAppearance.BackColor;
            SizeF HourSize = cache.Graphics.MeasureString("Hour", PopupViewInfo.PopupAppearance.Font);
            SizeF MinuteSize = cache.Graphics.MeasureString("Minute", PopupViewInfo.PopupAppearance.Font);

            PointF HourPoint = CalcLabelPoint(PopupViewInfo.CellsList[1], HourSize);
            PointF MinutePoint = CalcLabelPoint(PopupViewInfo.CellsList[13], MinuteSize);

            cache.Graphics.DrawString("Hour", PopupViewInfo.PopupAppearance.Font, drawPen.Brush, HourPoint);
            cache.Graphics.DrawString("Minute", PopupViewInfo.PopupAppearance.Font, drawPen.Brush, MinutePoint);

        }

        protected virtual void DrawCells(GraphicsCache cache)
        {
            foreach (CellViewInfo cellInfo in PopupViewInfo.CellsList)
            {
                drawPen.Color = PopupViewInfo.PopupAppearance.BackColor;
                if (cellInfo.isSelected)
                {
                    ObjectPainter.DrawObject(cache, SkinElementPainter.Default, GetCellElementInfo(cellInfo, cache, true));
                    drawPen.Color = PopupViewInfo.PopupAppearance.BackColor2;
                }
                if (cellInfo.isHotPointed)
                {
                    ObjectPainter.DrawObject(cache, SkinElementPainter.Default, GetCellElementInfo(cellInfo, cache, false));
                    drawPen.Color = PopupViewInfo.PopupAppearance.BackColor2;
                }
                cache.Graphics.DrawString(cellInfo.CellText, cellInfo.cellTextFont,drawPen.Brush, SetTextPoint(cellInfo, cache.Graphics));
            }
        }

        private PointF SetTextPoint(CellViewInfo cellInfo, Graphics g)
        {
            SizeF textSize = g.MeasureString(cellInfo.CellText, PopupViewInfo.PopupAppearance.Font);
            float x = cellInfo.cellBounds.X + cellInfo.cellBounds.Width - 3 - textSize.Width;
            float y = cellInfo.cellBounds.Y + cellInfo.cellBounds.Height - 1 - textSize.Height;

            return new PointF(x, y);
        }

        protected virtual SkinElementInfo GetCellElementInfo(CellViewInfo cell, GraphicsCache cache, bool selected)
        {
            Skin skin = CommonSkins.GetSkin(PopupViewInfo.ownerControl.OwnerEdit.LookAndFeel.ActiveLookAndFeel);
            SkinElementInfo skinInfo;
            if (selected)
                skinInfo = new SkinElementInfo(skin[CommonSkins.SkinSelection], cell.cellBounds);
            else
                skinInfo = new SkinElementInfo(skin[CommonSkins.SkinHighlightedItem], cell.cellBounds);            
            skinInfo.Cache = cache;
            return skinInfo;
        }

        private PointF CalcLabelPoint(CellViewInfo cell, SizeF textSize)
        {
            const int indent = 2;
            float x = cell.cellBounds.X + cell.cellBounds.Width / 2 - textSize.Width / 2;
            float y = cell.cellBounds.Y - textSize.Height - indent;
            return new PointF(x, y);
        }
    }
}
